---
title: "A1"
output:
  html_document:
    toc: true
date: "r Sys.Date()"
bibliography: A1.bib
---
[@lin2018apoe4]
[@durinck2009a]
[@durinck2005a]
[@davis2007a]
[@huber2015a]
[@r]
[@sanghi2021a]
[@law2016a]
[@robinson2010a]
[@mccarthy2012a]
[@chen2016a]
[@mccarthy2012a]
[@morgan2021a]

# 1. Preperation
Download the required packages
```{r, message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    
if (!requireNamespace("GEOmetadb", quietly = TRUE))
    BiocManager::install("GEOmetadb")

if (!requireNamespace("GEOquery", quietly = TRUE))
    BiocManager::install("GEOquery")

if (!requireNamespace("knitr", quietly = TRUE))
    BiocManager::install("knitr")

if (!requireNamespace("edgeR", quietly = TRUE))
    BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
    BiocManager::install("biomaRt")

if (!requireNamespace("magrittr", quietly = TRUE))
    BiocManager::install("magrittr")

if (!requireNamespace("kableExtra", quietly = TRUE))
    BiocManager::install("kableExtra")

if(!file.exists('GEOmetadb.sqlite')) 
  GEOmetadb::getSQLiteFile()


library("BiocManager")
library("GEOmetadb")
library("GEOquery")
library("knitr")
library("edgeR")
library("biomaRt")
library("magrittr")
library("kableExtra")
```
<br>

# 2. Select an Expression Data Set 
The process of selecting the optimal dataset of interest are documented in the course journal
<br>

# 3. Clean the data and map to HUGO symbols

## 3.1 Download the expression data

Get the GEO description of the data set. 
```{r message=FALSE}
gse <- GEOquery::getGEO("GSE102956",GSEMatrix=FALSE)
knitr::kable(data.frame(head(GEOquery::Meta(gse))), format = "html")
```
\
Get platform info
```{r, message=FALSE}
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))
```
Platform title: `r current_gpl_info$title`\
Submission date: `r current_gpl_info$submission_date`\
Last update date: `r current_gpl_info$last_update_date`\
Organism: `r current_gpl_info$organism`\
\
Get the supplementary files.
```{r message=FALSE}
#if(!file.exists('GSE102956'))
sfiles <- GEOquery::getGEOSuppFiles('GSE102956')
fnames <- rownames(sfiles)
#There are 4 files 
fnames

```
\
I will choose the fourth file about the gene counts of neurons. 
```{r, message=FALSE}
expr_data = read.delim(fnames[4],header=TRUE, check.names = FALSE)
#Check what it looks like
knitr::kable(expr_data[1:10, 1:6], format = "html")
```
## 3.2 Assess data quality

How many genes have been measured.
```{r, message=FALSE}
dim(expr_data)
#It is at a reasonable level since there are 20000~25000 genes in human cells.
```
\
Show column names to see controls and mutants. 
```{r, message= FALSE}
colnames(expr_data)
#There are three controls (NEU_E3) and three test stimulation (NEU_E4) that are very reasonable and clear. 
```
\
Check duplication
```{r}
summarized_gene_counts <- sort(table(rownames(expr_data)),
                               decreasing = TRUE)
#Output gene duplication frequency
kable(summarized_gene_counts[which(summarized_gene_counts>1)[1:10]], format = 'html')
#There is no duplication. 
```
\
Number the three controls and tests neurons. 
```{r message=FALSE}
colnames(expr_data) <- c( "NEU_E3.1", "NEU_E3.2", "NEU_E3.3", "NEU_E4.1", "NEU_E4.2", "NEU_E4.3")
colnames(expr_data)
```
\
Define the groups
```{r message=FALSE}
samples <- data.frame(lapply(colnames(expr_data), 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(1,2)]}))
colnames(samples) <- colnames(expr_data)
rownames(samples) <- c("cell_type","cell_number")
samples <- data.frame(t(samples))
```

## 3.3 Cleaning 

Filter out the low counts
```{r message=FALSE}
#remove the low count genes through count per million function in edgR package.
#We want the data that has cpm value greater than one, and also since there are 
#three control and three stimulation, we want at least three counts of cpm > 1
cpms = edgeR::cpm(expr_data)
keep = rowSums(cpms > 1) >= 3
expr_data_filtered = expr_data[keep,]

#See the dim of filtered data.
dim(expr_data_filtered)
#13407, way smaller than 23710 before cleaning
```
## 3.4 Mapping 

Convert NCBI gene symbol to NCBI gene numbers (e.g. from LOC12345 to 12345)
```{r message=FALSE}
a <- grep("^LOC",rownames(expr_data_filtered))
genename <- rownames(expr_data_filtered[a,])
len <- length(rownames(expr_data_filtered[a,]))
for (i in 1:len){
  genename[i] <-  substring(genename[i], 4)
}
row.names(expr_data_filtered)[a] <- genename
```

Check all available marts
```{r message=FALSE}
listMarts()
```
\
Go with ensembl mart 
```{r message=FALSE}
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
```
\
Choose the human data set 
```{r message=FALSE}
#search for human
knitr::kable(head(datasets[grep(datasets$dataset, pattern = "sapiens"),]),format = "html")

ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
```
\
Look for the right filter 
```{r message=FALSE}
biomart_human_filters <- listFilters(ensembl)
kable(biomart_human_filters[grep(biomart_human_filters$name,pattern="gene_id"),],format="html") %>% 
  row_spec(3, background = "yellow")
```
\
Look for attribute 
```{r message=FALSE}
kable(searchAttributes(mart = ensembl, 'gene_id')[1:12,] , 
      format="html") %>%
  row_spec(c(1,11), background = "yellow")

kable(searchAttributes(mart = ensembl, 'hgnc_symb')[1:12,] , 
      format="html") %>%
  row_spec(c(1,11), background = "yellow")
```
\
Mapping the entrezgene_id to hgnc_id
```{r message=FALSE}
conversion_stash <- "gene_conversion.rds"
if (file.exists(conversion_stash)) {
  gene_conversion <- readRDS(conversion_stash)
} else {
  gene_conversion <- biomaRt::getBM(attributes = c("entrezgene_id", "hgnc_symbol"),
                                  filters = "entrezgene_id",
                                  values = rownames(expr_data_filtered),
                                  mart = ensembl)
  saveRDS(gene_conversion, conversion_stash)
}

#See how many gene IDs have been trasnalted to HGNC
length(gene_conversion$hgnc_symbol) 
#How many gene IDs does not translate to HGNC
len - length(gene_conversion$hgnc_symbol) 
```

# 4. Normalization 

## 4.1 Data Distribution

Boxplot
```{r message=FALSE, warning=FALSE}
data2plot <- log2(edgeR::cpm(expr_data_filtered))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "APOE RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```
\
Density Plot 
```{r message=FALSE}
counts_density <- apply(log2(edgeR::cpm(expr_data_filtered)), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(counts_density)) {
      xlim <- range(c(xlim, counts_density[[i]]$x)); 
      ylim <- range(c(ylim, counts_density[[i]]$y))
    }
    cols <- rainbow(length(counts_density))
    ltys <- rep(1, length(counts_density))
    #plot the first density plot to initialize the plot
    plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="", cex.lab = 0.85)
    #plot each line
    for (i in 1:length(counts_density)) 
      lines(counts_density[[i]], col=cols[i], lty=ltys[i])
    #create legend
    legend("topright", colnames(data2plot),  
           col=cols, lty=ltys, cex=0.75, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90")
```

## 4.2 Normalize with TMM 

MA plot for my data,\ 
comparing the first control neuron (NEU_E3.1) and second control neuron (NEU_E3.2).
comparing the first tests neuron (NEU_E4.1) and second tests neuron (NEU_E4.2).
comparing the first control neuron (NEU_E3.1) and first tests neuron (NEU_E4.1).
```{r message=FALSE}
plotMA(log2(expr_data_filtered[,c(1,2)]), ylab="M - ratio log expression", 
       main="NEU_E3.1 vs NEU_E3,2 two controls example")
plotMA(log2(expr_data_filtered[,c(4,5)]), ylab="M - ratio log expression", 
       main="NEU_E4.1 vs NEU_E4.2  two tests example")
plotMA(log2(expr_data_filtered[,c(1,4)]), ylab="M - ratio log expression", 
       main="NEU_E3.1 vs NEU_E4.1  example")
```
\
Create edgerR container for RNASeq count data 
```{r message=FALSE}
filtered_data_matrix <- as.matrix(expr_data_filtered)
d = edgeR::DGEList(counts=filtered_data_matrix, group=samples$cell_type)
d
```
\
Calculate the normolization factor 
```{r message = FALSE}
d = edgeR::calcNormFactors(d)
d
```
\
Normalized data 
```{r message=FALSE}
normalized_counts <- cpm(d)
```
\
Side by side box plot of original data and normalized data. 
```{r message=FALSE, warning=FALSE}
data1plot <- log2(edgeR::cpm(expr_data_filtered))
data2plot <- log2(edgeR::cpm(normalized_counts))
par(mfrow = c(1, 2)) 
boxplot(data1plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Original Samples")
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Normalized Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```
\
MDS plot
```{r message=FALSE}
plotMDS(d, labels=rownames(samples),
  col = c("darkgreen","blue")[factor(samples$cell_type)])
```
\

## 4.3 Diespersion 

Calculate dispersion using estimateDisp fucntion. 
```{r message=FALSE}
model_design <- model.matrix(~samples$cell_number 
                             + samples$cell_type+0)
d <- estimateDisp(d, model_design)
```
\
Graphing the dispersion-squared BCV
```{r message=FALSE}
plotBCV(d,col.tagwise = "black",col.common = "red",)
```
\
Visual representation of the mean-variance relationship
```{r message=FALSE}
edgeR::plotMeanVar(d, show.raw.vars = TRUE, show.tagwise.vars=TRUE, 
            show.ave.raw.vars = TRUE,  
            NBline=TRUE,
            show.binned.common.disp.vars = TRUE)
```
\

# 5. Interpretation 

1. What are the control and test conditions of the dataset?\
The controls of my data sets are three iPSC-derived neurons with apolipoprotein E3 (APOE3) variant.\
The tests are three iPSC-derived neurons with apolipoprotein E4 (APOE4) variant.\

2. Why is the dataset of interest to you?\
I am very interested in neuroscience and particularly neurodegenerative disease like Alzheimer's disease. This paper studies the APOE4 variant effects on Amyloid beta piptide, which makes things evern more interesting, because there is a debate going on about the amyloid scandal.  \

3. Were there expression values that were not unique for specific genes? How did you handle these?\
There were expression values that are not unique for some genes across different neurons. However, since we have three similar iPSC-derived control neurons, and similar test neurons. There would be no surprise that we have same gene expression values across neurons. I would not remove them, since they also provide valuable information about how controls and tests are different within themselves. \

4. Were there expression values that could not be mapped to current HUGO symbols?\
In my data sets, I do not have ensembl gene Id. Instead, I have NCBI gene IDs mixed with HUGO symbols in the same column.
As shown in the code above, 142 of gene IDs can be mapped into HUGO symbols, 109 genes cannot be mapped to HUGO symbols, which are non-characterized genes that only has NCBI gene IDS\

5. How many outliers were removed?\
Originally, there were 23710 genes in total. After we do cmp (count per million)>1 and rowcount >=3 to remove the low counts, there are 13407 genes left. 23710 - 13407 = 10303 outliers are removed.\

6. How did you handle replicates?\
There is no replicates in my data, which is checked in the code above.\

7. What is the final coverage of your dataset?\
The final dataset has 13407 unique genes with identifiers.  Percentage: 13407/23710 *100% = 56.55% 
\

# Reference 


